cmd/kwild/server/build.go

```diff
@@ -567,14 +570,50 @@ func buildSigner(d *coreDependencies) *auth.Ed25519Signer {
 func buildDB(d *coreDependencies, closer *closeFuncs) *pg.DB {
 	// Check if the database is supposed to be restored from the snapshot
 	// If yes, restore the database from the snapshot
-	restoreDB(d)
+	fromSnapshot := restoreDB(d)
 
 	db, err := d.dbOpener(d.ctx, d.cfg.AppCfg.DBName, 24)
 	if err != nil {
 		failBuild(err, "kwild database open failed")
 	}
 	closer.addCloser(db.Close, "closing main DB")
 
+	// DB has been restored from the snapshot. Need to do some cleanup
+	// - Remove all the pending migration, changeset, validator join and validator remove resolutions
+	// - Fix the expiry heights of all the pending resolutions
+	// (how to handle this for offline migrations? we have no way to know the last height of the old chain)
+	if fromSnapshot {
+		tx, err := db.BeginTx(d.ctx)
+		if err != nil {
+			failBuild(err, "failed to start transaction for cleanup after snapshot restore")
+		}
+		defer tx.Rollback(d.ctx)
+
+		resolutionTypes := []string{
+			migrations.StartMigrationEventType,
+			migrations.ChangesetMigrationEventType,
+			voting.ValidatorJoinEventType,
+			voting.ValidatorRemoveEventType,
+		}
+		if err = voting.DeleteResolutionsByType(d.ctx, tx, resolutionTypes); err != nil {
```

Comment:
It doesn't bother me too much, but we're doing a fair bit more than we probably should in cmd/kwild/server/build.go.  Is all these easily encapsulated in another package perhaps?

Reply:
Let me move this to voting